### PR TITLE
HDDS-9063. OM - Recon unable to obtain Ozone Manager DB Snapshot

### PR DESCRIPTION
**Pull request by devmadhuu**
## What changes were proposed in this pull request?

Recon is unable to load data from the OM because the DB sync is failing due to path issues:
```
2023-07-20 23:44:50,686 ERROR [pool-27-thread-1]-org.apache.hadoop.ozone.recon.spi.impl.OzoneManagerServiceProviderImpl: Unable to obtain Ozone Manager DB Snapshot.
java.io.FileNotFoundException: /var/lib/hadoop-ozone/recon/om/data/om.snapshot.db_1689896689992/ints/om.db_checkpoint_1689896690655/000103.sst (No such file or directory)
        at java.io.FileOutputStream.open0(Native Method)
        at java.io.FileOutputStream.open(FileOutputStream.java:270)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
        at org.apache.hadoop.ozone.recon.ReconUtils.untarCheckpointFile(ReconUtils.java:198)
```

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9063

## How was this patch tested?

This patch was tested using Junit test case as well after deploying patch in cluster.

### CODE REVIEW

**Comment by devmadhuu** [comment]
@GeorgeJahad Thanks for providing info and help, Pls review the patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java [hunk 1]
```diff
@@ -389,6 +395,19 @@ private void writeFilesToArchive(Set<Path> copyFiles,
...
 
+  @NotNull
+  private static Path getVerifiedCheckPointPath(Path checkpointLocation) {
+    Path metaDirPath = checkpointLocation;
+    // This check is done to take care of findbug else below getParent()
+    // should not be null.
+    Path locationParent = checkpointLocation.getParent();
+    if (null != locationParent) {
+      Path parent = locationParent.getParent();
+      metaDirPath = parent != null ? parent : locationParent;
```
Comment (GeorgeJahad) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java:406]: This should throw a null pointer exception.  switching it back to locationParent will cause a more confusing error.

**Comment by GeorgeJahad** [comment]
With the exception of the npe comment above, this looks good to me.  FYI @smengcl I wrote most of this patch after @devmadhuu  pointed out the problem to me from a real cluster.  He then kindly tested it to confirm the fix, created the PR and fixed the findbugs issues.

Thanks for all the help Devesh!

Given that I wrote most of it, we should probably get one more person to approve it, but it is a relatively safe, and important fix, so I will merge it tomorrow morning if no one else has time to review it before then.

**Comment by GeorgeJahad** [comment]
Also, to summarize the issue @devmadhuu found, the bootstrapper was using this method:  ServerUtils.getOzoneMetaDirPath() to determine the path of the metadir.  It turns out that in some configurations that method doesn't seem to work as expected.

So I just replaced it with a call to "checkpointLocation.getParent().getParent()".

Functionally, that is all this PR does.  The other changes just deal with findbug issues or add error checking.

**Comment by errose28** [comment]
> Also, to summarize the issue @devmadhuu found, the bootstrapper was using this method: ServerUtils.getOzoneMetaDirPath() to determine the path of the metadir. It turns out that in some configurations that method doesn't seem to work as expected.

Yeah I think most/all of our docker tests don't set om.db.dirs and just rely on ozone.metadata.dirs as the fallback. In a real deployment om.db.dirs would be set and that is where all the RocksDB info would be. `OMStorage#getOmDbDir` can be used instead of `ServerUtils#getOzoneMetaDirPath` to account for this. I think the way it is done here is fine though, since the location is already stored in the `DBCheckpoint` object.

**Review by errose28 — COMMENTED** [review]
Just tested this on a real (non-docker) deployment and it fixed the issue for me. I think we can merge once George's comment is resolved.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java [hunk 2]
```diff
@@ -350,13 +352,13 @@ private boolean includeSnapshotData(HttpServletRequest request) {
...
 
-  private void writeFilesToArchive(Set<Path> copyFiles,
-                                   Map<Path, Path> hardLinkFiles,
-                                   ArchiveOutputStream archiveOutputStream,
-                                   boolean completed)
+  private void writeFilesToArchive(
+      Set<Path> copyFiles,
+      Map<Path, Path> hardLinkFiles,
+      ArchiveOutputStream archiveOutputStream,
+      boolean completed, Path checkpointLocation)
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java:359]: nit
```suggestion
      boolean completed,
      Path checkpointLocation)
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java (hunk 1 above)
Comment (devmadhuu) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java]: @GeorgeJahad - Thanks for your review. I have taken care of the comment. Pls re-review.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java (hunk 2 above)
Comment (devmadhuu) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java]: @smengcl  - Thanks for your review. Taken care.

**Review by sumitagrawl — APPROVED** [review]
@devmadhuu thanks for working over this, LGTM +1

**Pull request merged**